go/worker/common/p2p/rpc/peermgmt.go

```diff
@@ -0,0 +1,291 @@
+package rpc
+
+import (
+	cryptorand "crypto/rand"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	core "github.com/libp2p/go-libp2p-core"
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/mathrand"
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+)
+
+const (
+	SuccessConnManagerPeerTagValue = 20
+	ShuffledBestPeerCount          = 5
+
+	// newPeerScoreMultiplier is the score multiplier for new peers for which we don't yet have any
+	// historical measurements.
+	newPeerScoreMultiplier = 0.9
+)
+
+// Inverse alpha (1/alpha) values for computing the exponential moving average of latencies used for
+// peer scoring. Split into peer-local and global EMAs.
+const (
+	peerInvAlpha   = 10
+	globalInvAlpha = 25
+)
+
+// PeerManager is an interface for keeping track of peer statistics in order to guide peer selection
+// when performing RPC requests.
+type PeerManager interface {
+	// AddPeer tries to add the given peer to the peer manager.
+	//
+	// Peer is only added in case it supports the specified protocol.
+	AddPeer(peerID core.PeerID)
+
+	// RemovePeer unconditionally removes the peer from the peer manager.
+	RemovePeer(peerID core.PeerID)
+
+	// RecordSuccess records a successful protocol interaction with the given peer.
+	RecordSuccess(peerID core.PeerID, latency time.Duration)
+
+	// RecordFailure records an unsuccessful protocol interaction with the given peer.
+	RecordFailure(peerID core.PeerID, latency time.Duration)
+
+	// RecordBadPeer records a malicious protocol interaction with the given peer.
+	//
+	// The peer will be ignored during peer selection.
+	RecordBadPeer(peerID core.PeerID)
+
+	// GetBestPeers returns a set of peers sorted by the probability that they will be able to
+	// answer our requests the fastest with some randomization.
+	GetBestPeers() []core.PeerID
+}
+
+type peerStats struct {
+	successes         int
+	failures          int
+	avgRequestLatency time.Duration
+}
+
+func (ps *peerStats) getScore(avgRequestLatency time.Duration) float64 {
+	if ps.successes+ps.failures > 0 {
+		// We have some history for this peer.
+		failRate := float64(ps.failures) / float64(ps.failures+ps.successes)
+		return float64(ps.avgRequestLatency) + failRate*float64(avgRequestLatency)
```

Comment:
So a peer failing all requests quickly, will have a similar (slightly worse) score than an average node returning all correct results? Hm, I guess that does make sense, since if the peer is failing quickly that's not really all that bad, as not a lot of time is spent on it.

Maybe the penalty part for failing could have an additional constant factor? But I have no reason to suggest anything other than `1`.